### DISCUSSION TITLE
Live stream automatic reconnect

### DISCUSSION THREAD

**Discussion by arnean** [gopro/labs#691] [Feature Requests]
Would it be possible to implement automatic reconnect feature for live streaming? 

I'm streaming adventure races - you are running or cycling or kayaking and if the stream disconnects you have to find the QR code printout and show it to the camera - and it's not easy, especially during the night - camera doesn't see the QR code - you have to illuminate it with your headlamp and the camera is super picky when the lightning is not perfect. QR recognition algorithm is not very good IMO. I managed to activate the streaming twice during this night race. Tried many-many times. It is really very upsetting. Stopped, changed the light to more uniform - nothing. 

It would be perfect to be able to start streaming with button press. If this is not possible then unlimited reconnection would somewhat alleviate the problem. Perhaps add option for this. 

Right now looking for other cameras that have button control and automatic reconnect.

**Comment by dnewman-gpsw** [comment]
GoPro does automatically reconnect, unless the network is gone for a long time. I used Livestreaming from an 8 hour endurance race, other than dropouts at some corners, Livestreaming continued.  

With an more unreliable network, you could simple have the camera start Livestreaming at boot, so then you only need to boot/reboot to continue. It might be possible to test the record status, to automatically restart the stream, although I haven't tried that.

**Comment by arnean** [comment]
Thank you very much for your answer and sorry for being too emotional on my first message. I re-read the documentation carefully. I did not pay attention to BOOT command but that would solve my problem nicely I think. If I understand it correctly I should show two QR codes to camera: 

!SAVEboot="!W!GMC"
!MBOOT="!Lboot"

And then every time I switch on the camera it will connect to Wifi and start streaming. And I can delete the boot file from SD to revert back to normal operation.

I didn't get the last sentence though: "It might be possible to test the record status, to automatically restart the stream, although I haven't tried that." Can you please clarify?

I looked at the SD card for log files. There was qrlog.txt file that had some info, but not too much. 

[2023-12-29 19:24:00] Labs:Main Start 
[2023-12-29 19:24:00] Labs:Initialize 
[2023-12-29 19:25:06] Read QR:!W!GMC 
[2023-12-29 19:25:18] joining network:aip 
[2023-12-29 19:25:22] LS State 1:20 
[2023-12-29 19:25:26] LS State 1:17 
[2023-12-29 19:25:28] LS State 2:18 
[2023-12-29 19:25:32] LS Setup Complete:LIVE-STREAMING NOW! 
[2023-12-29 19:47:28] Labs:Main Start 
[2023-12-29 19:47:28] Labs:Initialize 
[2023-12-29 19:48:43] Read QR:!W!GMC 
[2023-12-29 19:48:52] joining network:aip 
[2023-12-29 19:48:56] LS State 1:20 
[2023-12-29 19:48:56] LS State 1:17 
[2023-12-29 19:48:58] LS State 2:18 
[2023-12-29 19:49:02] LS Setup Complete:LIVE-STREAMING NOW! 
[2023-12-29 20:15:40] Labs:Main Start 
[2023-12-29 20:15:40] Labs:Initialize 
[2023-12-29 20:28:58] Labs:Main Start 
[2023-12-29 20:28:58] Labs:Initialize 
[2023-12-29 20:29:14] Read QR:!W!GMC 
[2023-12-29 20:29:23] joining network:aip 
[2023-12-29 20:29:26] LS State 1:17 
[2023-12-29 20:29:28] LS State 2:18 
[2023-12-29 20:29:31] LS Setup Complete:LIVE-STREAMING NOW! 
[2023-12-29 20:38:19] Labs:Main Start 
[2023-12-29 20:38:19] Labs:Initialize 
[2023-12-29 20:43:26] Labs:Main Start 
[2023-12-29 20:43:26] Labs:Initialize 
[2023-12-29 21:20:48] Labs:Main Start 
[2023-12-29 21:20:49] Labs:Initialize 
[2023-12-29 21:29:01] Labs:Main Start 

There is no indication why streaming stopped. Or any info about network disconnects. Is it possible to somehow increase the logging level? I will try to experiment some more before next race.

Thanks again for quick answer.

**Reply by dnewman-gpsw** [reply]
The log reports anything that Labs does, not what internal operation might be happening. However using !MDBGL=1 will get you more from Labs.  This appears to be an older camera, as it doesn't list the firmware version. Which model is this?

In the above, when did LS stop, at what time(s) approximately?

**Comment by arnean** [comment]
I was streaming to Twitch. First transmission in Twitch is 12:04 long and 12:34 on SD card. So it ended between 19:37:34 and 19:38:06. Second transmission in Twitch is 11:59 and 12:29 on SD card. End between 20:01:01 and 20:01:31. Third one is 2:17 in Twitch and 3:00 on SD card. 

Camera is GoPro 10. Latest firmware 1.50.70. 

I will try the DBGL command.

**Reply by dnewman-gpsw** [reply]
While I can only simulate a connection lost by walking out of WiFi range (you are likely have a reliable WiFi signal, but your 4/5G connection is being lost) this allows for continued reconnections.

first set you camera up to LS.  oW1mVr1080!W  <- WiFi on video 1080p and connect to established network.

Then run:   <r0!GLC>r0!5N+!R~!5R

loop:
if(!record) // <r0
  !GLC // Go Live with Large (1080p) and capture a local file
if(recording) // >r0
   Sleep 5s,  Repeat  // !5N+!R
else
   Repeat in 5s  // !5R (this one has the cancel option)

This should work on 1.50.70. I did test it on HERO10, but using a slightly new build (not yet released.)

**Comment by arnean** [comment]
It seems that the published firmware does not support the camera status conditionals. This is also what the documentation says:

"Conditionals Based on Camera Status (HERO11 & 12)"

This is what I got in the log:

[2024-01-01 16:31:01] Read QR:<r0!GLC>r0!5N+!R~!5R 
[2024-01-01 16:31:01] Use CMD:<r0!GLC>r0!5N+!R~!5R 
[2024-01-01 16:31:01] CONDITION:1 
[2024-01-01 16:31:01] true 16:31 < 664:18:<r0!GLC>r0!5N+!R~!5R 
[2024-01-01 16:31:01] unknown token:
 
[2024-01-01 16:31:01] RES token:0 
[2024-01-01 16:31:01] vmode:done 
[2024-01-01 16:31:01] presets_edit:done 
[2024-01-01 16:31:01] ACTION:!5N+!R~!5R 
[2024-01-01 16:31:07] delay:end 
[2024-01-01 16:31:08] presets_edit:done 
[2024-01-01 16:31:08] ACTION:!R~!5R 
[2024-01-01 16:31:09] GP2 Temp.:35 

BTW I just discovered that !W is not needed before !G. Just !G will join the Wifi network. It seems that connecting and disconnecting is based on if the connection is needed or not.

**Reply by dnewman-gpsw** [reply]
You are correct.  The more advanced conditions haven't made it back to HERO10 (yet).  If you want to try an early test build of the these features on HERO10  https://www.dropbox.com/scl/fi/01jzqvijbvfz97lgkf2in/HERO10LabsV15074-23Decy21.zip?rlkey=xvk4k6n3ob7ckmighvp4wtub1&dl=0  This is not an official release, and not tested. It is stock 1.50.7x with the newer Labs command set.

**Reply by arnean** [reply]
Thanks for help. Unfortunately it failed to install. I tried thee times, first time over the latest Labs version, then installed the stock firmware and it still failed. I noticed that the FWUPD.TXT file changed on SD. It's the new contents:

```
# Load sequence
CAMERA:-4b4
# Camera upgrade rules file
OPTIONS:0
```

**Comment by arnean** [comment]
I will need to do some live streaming again and checked out what has changed. Found a new Labs firmware and installed it. First of all it seems that the streaming code itself is now handling the reconnects. I also tried your reconnect script and this also works. So it's like a double defence. I will do some field testing but it seems to be much more robust now. Thank you very much!

**Comment by Glacierwulf** [comment]
Just stopping by to say Thanks!
I ended up using !12:45!W<r0!GLC>r0!10N+!R~!10 for a 12:45pm start.
 I found that a non-number is needed between the time and the < to get things to work
Hero Black 11 for the record

**Reply by dnewman-gpsw** [reply]
That is not a fully valid command. It is not completely clear what you want it to do.

!12:45 is invalid.  !12:45N or !12:45S, etc., you need to set the action for the sleep command.  I expect the poor formatting is prevent the !R (repeat) from canceling are working live stream.

**Reply by Glacierwulf** [reply]
i see what i did now
i think !12:45W<r0!GLC>r0!10N+!R~!10 is more proper (yes i tested before i responded)
Thanks for the help

**Comment by Anammox** [comment]
This sounds like what I'm looking for but I'm confused. Sorry, I'm a noob.

I have a GoPro 11 and stream to rtmp server. When I go through a bad spot it ends the stream. Probably 1 or 2 minutes of poor or absent cell service. 

On not trying to record - just keep the rtmp stream live or trying to reconnect.

Does the labs firmware resolve this by attempting more reconnects? Is there a way to set parameters for the attempted reconnects, like number of attempts or duration of time? Does it work by default or do I need custom code?

TIA, much appreciated.

**Comment by LukePittman** [comment]
Another bump for this.  I have an event where a camera will go in and out of wifi signal through each competitors run through the course.  It could be disconnected for 3 or 4 minutes, or longer, but I need it to continually attempt to reconnect.  If it tries until the battery dies, that's fine.

Anyone have any further ideas here?

**Reply by dnewman-gpsw** [reply]
And the script for that posted in this thread (Dec 30, 2023) doesn't work?

**Reply by LukePittman** [reply]
Wow, I can't believe I didn't see that.  Thank you for calling me out, lol.  Yes, as a matter of fact, it works excellent.
